astropy/modeling/powerlaws.py

```diff
@@ -507,3 +507,115 @@ def input_units(self):
     def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
         return {'x_0': inputs_unit[self.inputs[0]],
                 'amplitude': outputs_unit[self.outputs[0]]}
+
+
+class Schechter1D(Fittable1DModel):
+    r"""
+    Schechter luminosity function (`Schechter 1976
+    <https://ui.adsabs.harvard.edu/abs/1976ApJ...203..297S/abstract>`_),
+    parameterized in terms of magnitudes.
+
+    Parameters
+    ----------
+    phi_star : float
+        The normalization factor in units of number density.
+
+    m_star : float
+        The characteristic magnitude where the power-law form of the
+        function cuts off. Must not have units.
+
+    alpha : float
+        The power law index, also known as the faint-end slope. Must not
+        have units.
+
+    See Also
+    --------
+    PowerLaw1D, ExponentialCutoffPowerLaw1D, BrokenPowerLaw1D
+
+    Notes
+    -----
+    Model formula (with :math:`\phi^{*}` for ``phi_star``, :math:`M^{*}`
+    for ``m_star``, and :math:`\alpha` for ``alpha``):
+
+    .. math::
+
+        n(M) \ dM = (0.4 \ln 10) \ \phi^{*} \
+            [{10^{0.4 (M^{*} - M)}}]^{\alpha + 1} \
+            \exp{[-10^{0.4 (M^{*} - M)}]} \ dM
+
+    ``phi_star`` is the normalization factor in units of number density.
+    ``m_star`` is the characteristic magnitude where the power-law form
+    of the function cuts off into the exponential form. ``alpha`` is
+    the power-law index, defining the faint-end slope of the luminosity
+    function.
+
+    Examples
+    --------
+    .. plot::
+        :include-source:
+
+        from astropy.modeling.models import Schechter1D
+        import astropy.units as u
+        import matplotlib.pyplot as plt
+        import numpy as np
+
+        phi_star = 4.3e-4 * (u.Mpc ** -3)
+        m_star = -20.26
+        alpha = -1.98
+        model = Schechter1D(phi_star, m_star, alpha)
+        mag = np.linspace(-25, -17)
+
+        fig, ax = plt.subplots()
+        ax.plot(mag, model(mag))
+        ax.set_yscale('log')
+        ax.set_xlim(-22.6, -17)
+        ax.set_ylim(1.e-7, 1.e-2)
+        ax.set_xlabel('$M_{UV}$')
+        ax.set_ylabel('$\phi$ [mag$^{-1}$ Mpc$^{-3}]$')
+
+    References
+    ----------
+    .. [1] Schechter 1976; ApJ 203, 297
+           (https://ui.adsabs.harvard.edu/abs/1976ApJ...203..297S/abstract)
+
+    .. [2] `Luminosity function <https://en.wikipedia.org/wiki/Luminosity_function_(astronomy)>`_
+    """
+
+    phi_star = Parameter(default=1., description=('Normalization factor '
+                                                  'in units of number density'))
+    m_star = Parameter(default=-20., description='Characteristic magnitude')
+    alpha = Parameter(default=-1., description='Faint-end slope')
+
+    @staticmethod
+    def evaluate(mag, phi_star, m_star, alpha):
+        """Schechter luminosity function model function."""
+        factor = 10 ** (0.4 * (m_star - mag))
```

Comment:
@larrybradley you mentioned that the reason why `m_star` cannot have units is this due to this computation.

I suggest you break this up into something like:
```python
factor_exp = 0.4 * (m_star - mag)

if isinstance(factor_exp, Quantity):
    factor_exp = factor_exp.value

factor = 10 ** factor_exp
```
This should fix the units issue.

Reply:
Thanks, @WilliamJamieson.  I added this, including checks for magnitude units (to prevent, e.g., `m_star = 10 * u.km`).

Reply:
I am a little worried... what if someone defines `m_star` in ABmag and then pass in `mag` in STmag? Is that possible?

Reply:
I assumed that would raise an error, but in fact it does not!  That is not good.  So I need to add checks that `m_star` and `mag` have the same (mag) unit.

Reply:
I added a check that `m_star.unit == mag.unit`.